File: get_diff.R
# --------------------------------------------
# Author: Raul
# Date: 2025-08-31
# Script: get_diff.R
# Note: This script creates the function that outputs
#       the estimated difference in means
#       given a fit model
# --------------------------------------------



# -----------------------------
# Estimate difference in means
# -----------------------------
get_diff <- function(ghat_B, delta_B, ghat_A, delta_A, pi_hat, Y) {
  muhat_B <- ghat_B + (delta_B*(Y - ghat_B)/(pi_hat))/(mean(delta_B/pi_hat))
  muhat_A <- ghat_A + (delta_A*(Y - ghat_A)/(1-pi_hat))/(mean(delta_A/(1-pi_hat)))
  diff_means <- mean(muhat_B - muhat_A)
  diff_var <- var(muhat_B - muhat_A)
  o <-list(diff_means = diff_means, 
           muhat_B = muhat_B, 
           muhat_A = muhat_A,
           diff_var = diff_var,
           pval = 2*(1-pnorm(abs(diff_means)/sqrt(diff_var))))
}


#get_diff <- function(ghat_1, delta_1, ghat_0, delta_0, pi_hat, Y) {
#  muhat_1 <- mean(ghat_1 + (delta_1*(Y - ghat_1)/(pi_hat))/(mean(delta_1/pi_hat)))
#  muhat_0 <- mean(ghat_0 + (delta_0*(Y - ghat_0)/(1-pi_hat))/(mean(delta_0/(1-pi_hat))))
#  diff_means <- muhat_1 - muhat_0
#  o <-list(diff_means = diff_means, muhat_1 = muhat_1, muhat_0 = muhat_0)
#}

File: python_scripts/multi_stage/get_diff.py
# --------------------------------------------
# Author: Raul
# Date: 2025-01-08
# Script: get_diff.py
# Note: This script creates the function that outputs
#       the estimated difference in means
#       given a fit model
# --------------------------------------------

import numpy as np
from scipy.stats import norm

def get_diff(ghat_B, delta_B, ghat_A, delta_A, pi_hat, Y):
    """
    Estimate difference in means using doubly robust estimation
    
    Parameters:
    - ghat_B: predicted outcomes for treatment B
    - delta_B: indicator for treatment B
    - ghat_A: predicted outcomes for treatment A  
    - delta_A: indicator for treatment A
    - pi_hat: propensity scores
    - Y: observed outcomes
    
    Returns:
    - Dictionary with difference estimates and statistics
    """
    
    # Avoid division by zero
    pi_hat = np.clip(pi_hat, 1e-6, 1-1e-6)
    
    # Doubly robust estimators
    muhat_B = ghat_B + (delta_B * (Y - ghat_B) / pi_hat) / np.mean(delta_B / pi_hat)
    muhat_A = ghat_A + (delta_A * (Y - ghat_A) / (1 - pi_hat)) / np.mean(delta_A / (1 - pi_hat))
    
    # Difference in means
    diff_means = np.mean(muhat_B - muhat_A)
    
    # Variance estimation
    diff_var = np.var(muhat_B - muhat_A) / len(muhat_B)
    
    # P-value (two-sided test)
    pval = 2 * (1 - norm.cdf(abs(diff_means) / np.sqrt(diff_var)))
    
    return {
        'diff_means': diff_means,
        'muhat_B': muhat_B,
        'muhat_A': muhat_A,
        'diff_var': diff_var,
        'pval': pval
    }